### PR TITLE
feat(api): add vcs/repo/run/hatchery/organization/region/user/plugins events

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

**Comment by ovh-cds** [comment]
CDS Report build-cds#19764.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-cds#19765.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19765.0 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

**Comment by ovh-cds** [comment]
CDS Report build-cds#19766.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19766.0 ✘  
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✘  
  * Standalone Tests v2 ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19766.1 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

**Comment by ovh-cds** [comment]
CDS Report build-cds#19769.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19769.1 ✘  
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✘  
  * Standalone Tests v2 ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19769.2 ✘  
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✘  
  * Standalone Tests v2 ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19769.3 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

**Comment by ovh-cds** [comment]
CDS Report build-cds#19783.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-cds#19783.1 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19783.1 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

**Review by richardlt — APPROVED** [review]

**Review by yesnault — APPROVED** [review]

File: engine/api/event_v2/event_permission.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	}
+	if u != nil {
+		e.UserID = u.ID
+		e.Username = u.Username
+	}
+	publish(ctx, store, e)
+}
+
+func PublishPermissionUpdatedEvent(ctx context.Context, store cache.Store, previousPerm, perm sdk.RBAC, u *sdk.AuthentifiedUser) {
+	previousBts, _ := json.Marshal(previousPerm)
```
Comment (fsamin) [inline engine/api/event_v2/event_permission.go:27]: Are we sure we need the previous struct ?
In my ming we have it in the previous event

File: engine/api/event_v2/event_permission.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+func PublishPermissionCreateEvent(ctx context.Context, store cache.Store, perm sdk.RBAC, u *sdk.AuthentifiedUser) {
+	bts, _ := json.Marshal(perm)
+	e := sdk.EventV2{
+		ID:         sdk.UUID(),
+		Type:       sdk.EventPermissionCreated,
+		Permission: perm.Name,
+		Payload:    bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_permission.go:19]: is it possible to have a nil user ?

File: engine/api/event_v2/event_permission.go [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+func PublishPermissionDeleteEvent(ctx context.Context, store cache.Store, perm sdk.RBAC, u *sdk.AuthentifiedUser) {
+	bts, _ := json.Marshal(perm)
+	e := sdk.EventV2{
+		ID:         sdk.UUID(),
+		Type:       sdk.EventPermissionDeleted,
+		Permission: perm.Name,
+		Payload:    bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_permission.go:51]: is it possible to have a nil user ?

File: engine/api/event_v2/event_organization.go [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+func PublishOrganizationDeleteEvent(ctx context.Context, store cache.Store, org sdk.Organization, u *sdk.AuthentifiedUser) {
+	bts, _ := json.Marshal(org)
+	e := sdk.EventV2{
+		ID:           sdk.UUID(),
+		Organization: org.Name,
+		Type:         sdk.EventOrganizationDeleted,
+		Payload:      bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_organization.go:34]: is it possible to have a nil user ?

File: engine/api/event_v2/event_hatchery.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+func PublishHatcheryDeleteEvent(ctx context.Context, store cache.Store, h sdk.Hatchery, u *sdk.AuthentifiedUser) {
+	bts, _ := json.Marshal(h)
+	e := sdk.EventV2{
+		ID:       sdk.UUID(),
+		Hatchery: h.Name,
+		Type:     sdk.EventHatcheryDeleted,
+		Payload:  bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_hatchery.go:51]: is it possible to have a nil user ?

File: engine/api/event_v2/event_hatchery.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+	previousH, _ := json.Marshal(previousHatchery)
+	bts, _ := json.Marshal(h)
+	e := sdk.EventV2{
+		ID:       sdk.UUID(),
+		Hatchery: h.Name,
+		Type:     sdk.EventHatcheryUpdated,
+		Previous: previousH,
+		Payload:  bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_hatchery.go:36]: is it possible to have a nil user ?

File: engine/api/event_v2/event_hatchery.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+
+func PublishHatcheryCreateEvent(ctx context.Context, store cache.Store, h sdk.Hatchery, u *sdk.AuthentifiedUser) {
+	bts, _ := json.Marshal(h)
+	e := sdk.EventV2{
+		ID:       sdk.UUID(),
+		Hatchery: h.Name,
+		Type:     sdk.EventHatcheryCreated,
+		Payload:  bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_hatchery.go:19]: is it possible to have a nil user ?

File: engine/api/event_v2/event_entity.go [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+	e := sdk.EventV2{
+		ID:         sdk.UUID(),
+		ProjectKey: ent.ProjectKey,
+		VCSName:    vcsName,
+		Repository: repoName,
+		Entity:     ent.Name,
+		Type:       sdk.EventEntityDeleted,
+		Payload:    bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_entity.go:22]: if the user can be nil if some case, we should never set it.

event data must be reliable

File: engine/api/event_v2/event_analysis.go [hunk 9]
```diff
@@ -2,33 +2,44 @@ package event_v2
...
 		ID:         sdk.UUID(),
 		ProjectKey: a.ProjectKey,
 		VCSName:    vcsName,
 		Repository: repoName,
 		Status:     a.Status,
 		Type:       sdk.EventAnalysisStart,
-		Payload:    *a,
+		Payload:    bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_analysis.go:22]: if the user can be nil if some case, we should never set it.

event data must be reliable

File: engine/api/event_v2/event_plugin.go [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+
+func PublishPluginCreateEvent(ctx context.Context, store cache.Store, p sdk.GRPCPlugin, u *sdk.AuthentifiedUser) {
+	bts, _ := json.Marshal(p)
+	e := sdk.EventV2{
+		ID:      sdk.UUID(),
+		Type:    sdk.EventPluginCreated,
+		Plugin:  p.Name,
+		Payload: bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_plugin.go:19]: is it possible to have a nil user ?

File: engine/api/event_v2/event_plugin.go [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+		e.UserID = u.ID
+		e.Username = u.Username
+	}
+	publish(ctx, store, e)
+}
+
+func PublishPluginUpdateEvent(ctx context.Context, store cache.Store, pOld, p sdk.GRPCPlugin, u *sdk.AuthentifiedUser) {
+	previousBts, _ := json.Marshal(pOld)
```
Comment (fsamin) [inline engine/api/event_v2/event_plugin.go:27]: do we need the previous one ?

File: engine/api/event_v2/event_plugin.go [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+
+func PublishPluginDeleteEvent(ctx context.Context, store cache.Store, p sdk.GRPCPlugin, u *sdk.AuthentifiedUser) {
+	bts, _ := json.Marshal(p)
+	e := sdk.EventV2{
+		ID:      sdk.UUID(),
+		Type:    sdk.EventPluginDeleted,
+		Plugin:  p.Name,
+		Payload: bts,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_plugin.go:51]: is it possible to have a nil user ?

File: engine/api/event_v2/event_run.go [hunk 13]
```diff
@@ -2,12 +2,38 @@ package event_v2
...
 	"context"
+	"encoding/json"
 
 	"github.com/ovh/cds/engine/cache"
 	"github.com/ovh/cds/sdk"
 )
 
+func PublishRunJobRunResult(ctx context.Context, store cache.Store, eventType, vcsName, repoName string, rj sdk.V2WorkflowRunJob, rr sdk.V2WorkflowRunResult) {
+	bts, _ := json.Marshal(rr)
+	e := sdk.EventV2{
```
Comment (fsamin) [inline engine/api/event_v2/event_run.go:13]: in fact i'm not sure it's a good idea to share the same structure for all events...

**Comment by ovh-cds** [comment]
CDS Report build-cds#19804.0 ✘  
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✘  
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report build-cds#19805.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19805.0 ✘  
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✘

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19805.1 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

**Comment by ovh-cds** [comment]
CDS Report build-cds#19811.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19811.0 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

File: engine/api/event_v2/event_analysis.go [hunk 14]
```diff
@@ -2,33 +2,48 @@ package event_v2
...
+		},
 		VCSName:    vcsName,
 		Repository: repoName,
 		Status:     a.Status,
-		Type:       sdk.EventAnalysisStart,
-		Payload:    *a,
+	}
+
+	// No user if we came from hook. User is resolved later during analysis process.
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_analysis.go:26]: remove this

File: engine/api/event_v2/event_entity.go [hunk 15]
```diff
@@ -0,0 +1,29 @@
...
+			ID:         sdk.UUID(),
+			Type:       eventType,
+			Payload:    bts,
+			ProjectKey: ent.ProjectKey,
+		},
+		VCSName:    vcsName,
+		Repository: repoName,
+		Entity:     ent.Name,
+	}
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_entity.go:24]:
```suggestion
  // User is nil for deletion (entity deletion is initiated by CDS itself)
	if u != nil {
```

File: engine/api/event_v2/event_hatchery.go [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+	e := sdk.HatcheryEvent{
+		GlobalEventV2: sdk.GlobalEventV2{
+			ID:      sdk.UUID(),
+			Type:    eventType,
+			Payload: bts,
+		},
+		Hatchery: h.Name,
+	}
+
+	if u != nil {
```
Comment (fsamin) [inline engine/api/event_v2/event_hatchery.go:22]:
```suggestion
  // user is nil for update event, because the hatchery updates itself
	if u != nil {
```

**Review by fsamin — CHANGES_REQUESTED** [review]

**Review by fsamin — APPROVED** [review]

**Comment by ovh-cds** [comment]
CDS Report build-cds#19824.0 ✔ 
* Build
  * Build and Package All ✔ 
  * Build Debpacker ✔ 
  * Build SMTPMock ✔ 
  * Lint UI Translate ✔ 
  * Test CLI ✔ 
  * Test Engine ✔ 
  * Test SDK ✔ 
  * Test UI ✔

**Comment by ovh-cds** [comment]
CDS Report integration-cds#19824.0 ✔ 
* Integration
  * Docker Compose Tests ✔ 
  * Standalone Tests ✔ 
  * Standalone Tests v1 ✔ 
  * Standalone Tests v2 ✔

**Pull request merged**